lib/Parser.js

```diff
@@ -0,0 +1,788 @@
+        return [{ tag: 'StringContent' }, { tag: 'NoFlush' }];
+      } else {
+        return backslashError(char);
+      }
+
+    case 'EmptyStringMaybeTriple':
+      switch (char) {
+        case '"':
+          return [{ tag: 'MultilineString' }, { tag: 'NoFlush' }];
+        default:
+          return [
+            { tag: 'Initial', otherTokenChars: char },
+            { tag: 'FlushToken', token: { tag: 'String' } },
+          ];
+      }
+
+    case 'MultilineString':
+      switch (char) {
+        case '"':
+          return [{ tag: 'MultilineString"' }, { tag: 'NoFlush' }];
+        case '\\':
+          return [{ tag: 'MultilineStringBackslash' }, { tag: 'NoFlush' }];
+        default:
+          return [{ tag: 'MultilineString' }, { tag: 'NoFlush' }];
+      }
+
+    case 'MultilineString"':
+      switch (char) {
+        case '"':
+          return [{ tag: 'MultilineString""' }, { tag: 'NoFlush' }];
+        case '\\':
+          return [{ tag: 'MultilineStringBackslash' }, { tag: 'NoFlush' }];
+        default:
+          return [{ tag: 'MultilineString' }, { tag: 'NoFlush' }];
+      }
+
+    case 'MultilineString""':
+      switch (char) {
+        case '"':
+          return [
+            { tag: 'Initial', otherTokenChars: '' },
+            { tag: 'FlushToken', token: { tag: 'String' } },
+          ];
+        case '\\':
+          return [{ tag: 'MultilineStringBackslash' }, { tag: 'NoFlush' }];
+        default:
+          return [{ tag: 'MultilineString' }, { tag: 'NoFlush' }];
+      }
+
+    case 'MultilineStringBackslash':
+      if (backslashableChars.has(char)) {
+        return [{ tag: 'MultilineString' }, { tag: 'NoFlush' }];
+      } else {
+        return backslashError(char);
+      }
+
+    default:
+      return unreachable(tokenizerState.tag);
+  }
+}
+
+const ParserState /*:
+  | {
+      tag: 'ModuleDeclaration',
+      lastToken: typeof ModuleDeclarationLastToken,
+    }
+  | {
+      tag: 'Rest',
+      lastToken: typeof RestLastToken,
+    } */ = { tag: 'ModuleDeclaration', lastToken: 'Nothing' };
+void ParserState;
+
+const ModuleDeclarationLastToken /*:
+  | 'Nothing'
+  | 'NewChunk'
+  | 'port/effect'
+  | 'module'
+  | 'ModuleName'
+  | 'ModuleName.'
+  | 'exposing'
+  | 'exposing('
+  | 'exposing..'
+  | 'LowerName'
+  | 'UpperName'
+  | 'UpperName('
+  | 'UpperName..'
+  | 'UpperName)'
+  | ',' */ = 'Nothing';
+void ModuleDeclarationLastToken;
+
+function parseModuleDeclaration(
+  token /*: typeof Token */,
+  lastToken /*: typeof ModuleDeclarationLastToken */
+) /*:
+  | typeof ModuleDeclarationLastToken
+  | 'StopParsing'
+  | 'NextParserState'
+  | SyntaxError */ {
+  switch (lastToken) {
+    case 'Nothing':
+      if (token.tag === 'NewChunk') {
+        return 'NewChunk';
+      }
+      return expected('a new chunk', token);
+
+    case 'NewChunk':
+      if (token.tag === 'Other') {
+        switch (token.value) {
+          case 'port':
+          case 'effect':
```

Comment:
So this tries to support `effect module` but I just learned that those look like this:

```elm
effect module ModuleName where { subscription = MySub } exposing (myVal)
```

So to support that for real there’s additional syntax to parse (`where { subscription = MySub }`). Only blessed modules can use `effect module`, right? Do we know if any of them uses elm-test and needs elm-test to be able to parse the `effect module` files to find all tests?

Reply:
The conclusion is:

Only packages can have `effect module` (and on top of that, only `elm/` and `elm-explorations/` packages), inside `src/`. Packages can only have tests inside `src/` if they depend directly on elm-explorations/test, which isn’t practical. So there’s no need to support effect modules. Instead we exit elm-test if we encounter one.

I’m going to require package tests to be in `tests/` as a separate commit.